### PR TITLE
Fix word_embeddings match for deepspeed wrapped model

### PR DESCRIPTION
**Pull request by mayank31398**
Fix for:
https://github.com/huggingface/transformers/pull/26623
https://github.com/huggingface/peft/issues/996

### CODE REVIEW

**Review by BenjaminBossan — CHANGES_REQUESTED** [review]
Thanks for providing the fix.

Could we make a change to increase the clarity of the code? Otherwise, it will be quite hard to understand a few months down the line. E.g., we could assign

`hasattr(value, "ds_shape") and value.ds_shape[0] == self.base_model.config.vocab_size`

to a variable with an informative name and then the check can be become

`if (value.shape[0] == self.base_model.config.vocab_size) or <var>`

Also, please add a comment with a link to the original issue.

**Comment by mayank31398** [comment]
addressed @BenjaminBossan

**Review by BenjaminBossan — APPROVED** [review]
That's much clearer, thanks for updating the PR. It LGTM, maybe @younesbelkada can do a final review too.

File: src/peft/peft_model.py [hunk 1]
```diff
@@ -315,7 +315,21 @@ def _setup_prompt_encoder(self, adapter_name: str):
...
+            is_deepspeed_distributed_tensor = hasattr(value, "ds_shape")
+
+            deepspeed_distributed_tensor_shape = None
+            if is_deepspeed_distributed_tensor:
+                deepspeed_distributed_tensor_shape = value.ds_shape
+
+            if value.shape[0] == self.base_model.config.vocab_size or (
+                is_deepspeed_distributed_tensor
+                and deepspeed_distributed_tensor_shape[0] == self.base_model.config.vocab_size
+            ):
```
Comment (younesbelkada) [inline src/peft/peft_model.py:332]:
```suggestion
            deepspeed_distributed_tensor_shape = getattr(value, "ds_shape", None)

            if value.shape[0] == self.base_model.config.vocab_size or (
                deepspeed_distributed_tensor_shape is not None
                and deepspeed_distributed_tensor_shape[0] == self.base_model.config.vocab_size
            ):
```

**Review by younesbelkada — APPROVED** [review]
Thanks a lot @mayank31398 ! I left a comment that slightly refactors the logic - what do you think?

**Comment by mayank31398** [comment]
yes @younesbelkada , your change looks better. I have added the suggestion.

**Review by younesbelkada — APPROVED** [review]
Thanks again!

**Pull request merged**